### PR TITLE
[Consensus] soft bundle to be included in a block atomically

### PR DESCRIPTION
**Pull request by akichidis**
## Description 

Following the work of https://github.com/MystenLabs/sui/pull/19572 , this PR is ensuring that a soft bundle will atomically be included in a consensus block. That:
1. Strengthens the guarantees around the soft bundle sequencing to allow them processed in exact sequence and not break in multiple blocks.
2. allow us to simply retry transactions of GC'ed blocks as we just retry all the transactions of softbundle

Kept the solution simple for now and if a soft bundle doesn't fit to a block then it's cached and is added to the next block.

## Test plan 

CI

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: consensus/core/src/transaction.rs [hunk 1]
```diff
@@ -217,6 +203,8 @@ impl TransactionConsumer {
 pub struct TransactionClient {
     sender: Sender<TransactionsGuard>,
     max_transaction_size: u64,
+    max_transactions_in_block_bytes: u64,
+    max_transactions_in_block_size: u64,
```
Comment (mwtian) [inline consensus/core/src/transaction.rs:207]:
```suggestion
    max_transactions_in_block_count: u64,
```
optional: or use the original config field. `size` may confuse with bytes size.

File: consensus/core/src/transaction.rs [hunk 2]
```diff
@@ -261,19 +261,39 @@ impl TransactionClient {
...
+    /// Similar if transactions exceed `max_transactions_in_block_size` an error will be returned.
     pub(crate) async fn submit_no_wait(
         &self,
         transactions: Vec<Vec<u8>>,
     ) -> Result<oneshot::Receiver<(BlockRef, oneshot::Receiver<BlockStatus>)>, ClientError> {
         let (included_in_block_ack_send, included_in_block_ack_receive) = oneshot::channel();
+
+        let mut bundle_size = 0;
+
+        if transactions.len() as u64 > self.max_transactions_in_block_size {
```
Comment (mwtian) [inline consensus/core/src/transaction.rs:275]: In `next()`, `max_consumed_transactions_per_request` is compared against with `>=`, so this should use `>=` as well?

**Review by mwtian — APPROVED** [review]

File: consensus/core/src/transaction.rs (hunk 2 above)
Comment (akichidis) [reply consensus/core/src/transaction.rs:275]: Good point. Actually I'll change the `max_consumed_transactions_per_request` to use `>` instead. So maximum will always mean inclusive.

File: consensus/core/src/transaction.rs (hunk 1 above)
Comment (akichidis) [reply consensus/core/src/transaction.rs:207]: `count` sounds good 👍

**Pull request merged**